web/src/components/DandisetList.vue

```diff
@@ -69,50 +69,79 @@
   </v-list>
 </template>
 
-<script>
+<script lang="ts">
+import {
+  defineComponent, ref, computed, watch, PropType,
+} from '@vue/composition-api';
 import moment from 'moment';
 import filesize from 'filesize';
 
 import { getDandisetContact } from '@/utils';
 import toggles from '@/featureToggle';
 import { girderRest } from '@/rest';
 
-export default {
+type Dandiset = {};
+interface DandisetStats {
+  bytes: number;
+  folders: number;
+  items: number;
+}
+
+export default defineComponent({
+  name: 'DandisetList',
   props: {
     dandisets: {
       // Girder Items
-      type: Array,
+      type: Array as PropType<Dandiset[]>,
       required: true,
     },
   },
-  computed: {
-    items() {
-      return this.dandisets;
-    },
-    origin() {
-      const { name, params, query } = this.$route;
+  setup(props, ctx) {
+    // Will be replaced by `useRoute` if vue-router is upgraded to vue-router@next
+    // https://next.router.vuejs.org/api/#useroute
+    const route = ctx.root.$route;
+
+    const origin = computed(() => {
+      const { name, params, query } = route;
       return { name, params, query };
-    },
-  },
-  asyncComputed: {
-    async dandisetStats() {
+    });
+
+    const items = computed(() => props.dandisets);
+    const dandisetStats = ref<DandisetStats[] | null>(null);
+    async function fetchDandisetStats(dandisets: Dandiset[]) {
+      // Set back to null in case of failure
+      dandisetStats.value = null;
+
       if (toggles.DJANGO_API) {
-        return this.dandisets;
+        dandisetStats.value = dandisets as DandisetStats[];
       }
-      const { items } = this;
-      return Promise.all(items.map(async (item) => {
-        const { identifier } = item.meta.dandiset;
+
+      const res = await Promise.all(dandisets.map(async (dandiset: any) => {
+        const { identifier } = dandiset.meta.dandiset;
         const { data } = await girderRest.get(`/dandi/${identifier}/stats`);
         return data;
       }));
-    },
-  },
-  methods: {
-    filesize,
-    getDandisetContact,
-    formatDate(date) {
+
+      dandisetStats.value = res;
+    }
+
+    // Fetching dandiset stats must be done this way since we don't have access to asyncComputed
```

Comment:
This could be rewritten with [watchEffect](https://composition-api.vuejs.org/api.html#watcheffect), it's basically asyncComputed for composition API.

Reply:
You mean change the call from `watch` to `watchEffect`? I suppose I could, but since I prefer to use `watch` when possible, since I'm only watching a single dependency.

Reply:
I think you make a good observation, and it seems `watchEffect` is more suited for future changes (i.e. adding more properties to watch). However, I don't want to hold up this PR any longer. This would be an easy change in the future anyway.

Reply:
@AlmightyYakob, are you saying you agree with @dchiquit's recommendation here (even if you didn't include it in this PR)? In particular, I'm wondering why you prefer `watch()` just because it's a single dependency you're watching? Why not just make `fetchDandisets()` itself "reactive" by embedding it in a `watchEffect()`?

Reply:
The main reason I've preferred `watch` over `watchEffect` is that when it comes to actually needing to watch something for changes (instead of using it as a way to achieve async computed props), I prefer to have it clearly defined what's being watched. However, it seems like `watchEffect` is better suited for this specific task, so in the future I'll probably go to `watchEffect` first, and then `watch` if a particular task requires it. There are different applications that are better suited for each one. More docs about this can be found [here](https://composition-api.vuejs.org/api.html#watch).

Reply:
Thanks @AlmightyYakob. I've read those docs before and they don't really provide concrete guidance on when to use what. Your intuition seems good in general, but I think I like `watchEffect()` specifically for cases where you're trying to emulate a reactive response involving imperative actions, or otherwise trying to bolt on imperative stuff to Vue's reactive system. Things like asyncComputed, or triggering D3 rendering, qualify IMO.

Thanks for the discussion. This is useful for developing mental models for these newer features of Vue.